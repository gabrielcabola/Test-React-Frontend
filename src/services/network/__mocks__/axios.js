// ./__mocks__/axios.js
import mockAxios from 'jest-mock-axios';
export default mockAxios;

export const articles = {
  "articles": [
    {
      "manufacturer": "Yamaha",
      "model": "C40",
      "mainImage": {
        "fileName": "135317.jpg"
      }
    }]
};