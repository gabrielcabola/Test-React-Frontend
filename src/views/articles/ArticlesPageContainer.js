import React, { createContext, useEffect } from "react";
import { request } from 'services/network/Network.service';
import { ActionTypes } from 'store/articles/Articles.actions';
import { ArticleReducer, ArticleInitialState } from 'store/articles/Articles.reducers';
import Loader from "components/loader/Loader";
import ArticlesPage from "views/articles/ArticlesPage";

// Create Context Object
export const ArticlesContext = createContext();
// API  fetch
export const fetchArticles = () => request('GET', '/articles');

// Create a provider for components to consume and subscribe to changes
export default function ArticlesPageContainer(props) {
    
   const [state, dispatch] = React.useReducer(ArticleReducer, ArticleInitialState);

    // Fetch the Information from Backend
    useEffect(() => {
        const { loading } = state;
        if (loading === false) {
          dispatch({ type: ActionTypes.INIT });
          fetchArticles().then(({ status, articles }) => {
                  dispatch({
                    type: ActionTypes.COMPLETE,
                    payload: articles
                  });
                }).catch((error) =>
                
                    console.log(error)
                );
        }
    }, [state]);

    const ArticlesModule = (
        <ArticlesContext.Provider value={[state, dispatch]}>
          <ArticlesPage {...state} />
        </ArticlesContext.Provider>
    );
    
    return (state.completed === true) ? ArticlesModule : <Loader />;
};

