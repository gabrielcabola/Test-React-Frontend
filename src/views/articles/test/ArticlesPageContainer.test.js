import React from "react";
import renderer from 'react-test-renderer';
import ArticlesPageContainer from "../ArticlesPageContainer";
import { render, screen } from "@testing-library/react";
import { ActionTypes } from 'store/articles/Articles.actions';
import { ArticleReducer, ArticleInitialState } from 'store/articles/Articles.reducers';
import { articles } from 'services/network/__mocks__/axios';


/**
 * Component
 */
describe("<ArticlesPageContainer />", () => {
  
  beforeEach(() => {
    render(<ArticlesPageContainer />);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<ArticlesPageContainer />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  describe("when page is initialized", () => {
    it("then shows the Loader", () => {
      const Loader = document.getElementById('Loader');
      expect(Loader).toBeTruthy();
    });

    it('handles fetch articles', () => {
      expect(ArticleReducer(ArticleInitialState, { type: ActionTypes.INIT })).toEqual({
        ...ArticleInitialState,
        loading: true,
      });
    });
  });
  
/**
 * Reducers
 */
describe('ArticleReducer', () => {
    
    it('returns the initial state', () => {
      expect(ArticleReducer(undefined, {})).toEqual(ArticleInitialState);
    });
  
    it('handles article request', () => {
      expect(ArticleReducer(ArticleInitialState, { type: ActionTypes.INIT })).toEqual({
        ...ArticleInitialState,
        loading: true,
      });
    });
  
    it('handles article request successful', () => {
      expect(ArticleReducer(ArticleInitialState, {
        type: ActionTypes.COMPLETE,
        payload:articles,
      })).toEqual({
        ...ArticleInitialState,
        completed: true,
        loading: true,
        articles,
      });
    });
});

  /**
   * Actions and Dispatchs
   */

  /**
   * TODOS:
   * 
   * Test Action Types and ispatchs using:
   *  'redux-mock-store'
   *  'redux-thunk'
   * 
   */

});