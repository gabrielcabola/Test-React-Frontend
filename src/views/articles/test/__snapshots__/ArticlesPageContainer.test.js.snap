// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticlesPageContainer /> renders correctly 1`] = `
<div
  className="Loading"
>
  <div
    className="loadingio-spinner-rolling-8tt3fggefq8"
    id="Loader"
  >
    <div
      className="ldio-xpemuty8dbl"
    >
      <div />
    </div>
  </div>
</div>
`;
